guardian/cdk
# Add pattern for SNS lambdas

## What does this change?

This PR adds a pattern for lambda's triggered by an SNS topic. It also refactors some of the alarm-related code added in https://github.com/guardian/cdk/pull/186 to make it easier to share that code with this new pattern.

## Does this change require changes to existing projects or CDK CLI?

Existing projects using the SNS -> lambda architecture should be able to reduce the amount of boilerplate in their templates. You can see an example of this here: https://github.com/guardian/google-chat-bots/pull/152.

Due to the aforementioned refactor, some projects might need to make minor changes (e.g. fixing import statements) for the alarm-related code.

## How to test

1. Create a new stack based on this pattern. You should end up with a functional lambda triggered by SNS messages.
2. Upgrade existing stacks to use this pattern:
    1. A stack which defines its own SNS topic
    2. A stack which uses an external SNS topic

## How can we measure success?

* Projects already using this library (e.g. `sns-to-gchat`) should be able to simplify their infrastructure definition
* Projects using this architecture can configure sensible monitoring for a tiny amount of extra code/effort
* It should be faster to set up new projects correctly using this pattern

## Have we considered potential risks?

When migrating existing stacks, SNS Subscriptions and Invoke permissions will be replaced due to changing ids (i.e. we create new versions of these resources and delete the old ones). I am reasonably confident that the `Create` operations will run before the `Delete` operations, which should prevent any disruption to service. Although I have tested this, I'm struggling to find docs from AWS which explicitly state that this will always be the case, so there is a small risk that migrating to this pattern could introduce a very brief period of downtime.
